wlms/client.go

```diff
@@ -366,35 +404,49 @@ func (c *Client) Handle_LOGIN(server *Server, pkg *packet.Packet) CmdError {
 		return CriticalCmdPacketError{err.Error()}
 	}
 
-	if c.protocolVersion != 0 && c.protocolVersion != 1 {
+	// Check protocol version
+	if c.protocolVersion != 0 && c.protocolVersion != 2 {
 		return CriticalCmdPacketError{"UNSUPPORTED_PROTOCOL"}
 	}
 
-	if isRegisteredOnServer || c.protocolVersion == 1 {
+	if isRegisteredOnServer || c.protocolVersion == 2 {
 		nonce, err := pkg.ReadString()
 		if err != nil {
 			return CriticalCmdPacketError{err.Error()}
 		}
 		c.nonce = nonce
 	}
 
+	// Check if registered. If it is, check credentials. If invalid, abort.
 	if isRegisteredOnServer {
-		if server.HasClient(c.userName) != nil {
-			return CriticalCmdPacketError{"ALREADY_LOGGED_IN"}
-		}
 		if !server.UserDb().ContainsName(c.userName) {
 			return CriticalCmdPacketError{"WRONG_PASSWORD"}
 		}
 		if !server.UserDb().PasswordCorrect(c.userName, c.nonce) {
 			return CriticalCmdPacketError{"WRONG_PASSWORD"}
 		}
 		c.permissions = server.UserDb().Permissions(c.userName)
-	} else {
-		baseName := c.userName
-		for i := 1; server.UserDb().ContainsName(c.userName) || server.HasClient(c.userName) != nil; i++ {
-			c.userName = fmt.Sprintf("%s%d", baseName, i)
+	}
+	// Check for clients which are using the same nonce
+	c.replaceCandidates = server.FindClientsToReplace(c.nonce, c.userName)
+	if len(c.replaceCandidates) == 0 {
+		// Noone connected with our nonce
+		// TODO(Notabilis): Maybe do the ContainsName() check here case-insensitive?
+		//                  Having "Peter" and "peter" is quite strange. Same for HasClient().
```

Comment:
We'd need to make sure then that Widelands itself is also case-insensitive. Too much effort?

Reply:
Should be no problem implementing this, though I don't think we have to change anything ingame for this. Except maybe when we also want to be case-insensitive when using whisper messages.

Reply:
Just wondering: Are you waiting for me to implement this? I wouldn't have done it as part of this branch but of course I can do so.

I looked into the code and for internet gaming it should be possible to change this as a pure metaserver-side change. When we also want to have this in LAN games we have to change Widelands, too.
Two questions regarding the SQL user database:
- Is the username column of the auth_user table an index? In that case the performance would most likely drop when we perform case-insensitive queries. (Could probably be circumvented by adding an always lower-case column which could be indexed.)
- Are there any usernames registered which will break when we implement this?

Reply:
Yes, we can always fix this later. I was sort of hoping that SirVer would pitch in, but looks like he doesn't have time at the moment.

Maybe wait with merging until Wednesday, when the current tournament games have been played - just in case?

Reply:
I am not that familiar with the things you are working on but will try to answer some questions:

The username of auth_user is a field of type vchar. So this is not an index. 
A username "Kaputtnik" is a different user as "kaputtnik" (usernames are always case sensitive)

> Are there any usernames registered which will break when we implement this?

If you mean a case insensitive search: I didn't searched for usernames, but if have found none it would mean to implement a thing to forbid a username 'Kaputtnik' when a username 'kaputtnik' already exists. This has to be done before you implement a case insensitive search.

Reply:
Thanks for the review! We could merge this, but as far as I know SirVer has to deploy it manually anyway. So lets just wait until he has time for it and merge the related Widelands branch when he deploys the metaserver.

Also thanks for the information regarding the database server, I totally forgot about the registration process. So we have to update the website, the metaserver and Widelands itself when we want to add this. I will open a bug report / feature request on launchpad for implementing this in the future.